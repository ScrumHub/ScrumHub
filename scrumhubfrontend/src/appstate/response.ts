import { IError } from "./stateInterfaces";

export type APIResponse<T, K> = {
    message: string;
    successful: boolean;
    data: T | null;
    metadata: K | null;
  };
  
  export type RequestResponse<T, K> = {
    code: number;
    response: APIResponse<T, K>|{}|IError;
  };