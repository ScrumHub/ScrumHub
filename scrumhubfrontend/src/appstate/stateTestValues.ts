
import config from "../configuration/config";
import { initError, initPeopleList, initPerson, initBI, initBacklogItemList, initRepository, initRepositoryList, initSprint, initSprintList, initTaskList, initTask, initKeys, initLoadingKeys } from "./stateInitValues";
import { ILoginState, IBacklogItemList, IRepositoryList, ISprintList, IState, ITaskList } from "./stateInterfaces";
import { validateUri } from "./reducerUtilities";

export const testLoginState: ILoginState = {
  isLoggedIn: false,
  token: "",
  client_id: validateUri(process.env.REACT_APP_CLIENT_ID),
  redirect_uri: validateUri(process.env.REACT_APP_REDIRECT_URI),
  client_secret: validateUri(process.env.REACT_APP_CLIENT_SECRET),
  proxy_url: validateUri(process.env.REACT_APP_PROXY_URL)
}

export const initTestState: IState = {
  loading: false,
  error: initError,
  pbiPage: initBacklogItemList,
  sprintPage: initSprintList,
  taskPage: initTaskList,
  repositories: initRepositoryList,
  openSprint: initSprint,
  reposLastPage: false,
  reposRequireRefresh: false,
  productLastPage: false,
  productRequireRefresh: false,
  sprintLastPage: false,
  sprintRequireRefresh: false,
  taskLastPage: false,
  taskRequireRefresh: false,
  tasks: [],
  namedPBI: [],
  repoId: -1,
  people: initPeopleList,
  currentUser: initPerson,
  activeSprintNumber: -1,
  keys: initKeys,
  loadingKeys: initLoadingKeys,
  loginState: testLoginState,
  changedRepo: ""
};

export const testRepositoryList: IRepositoryList = {
  pageNumber: 1,
  pageCount: 1,
  pageSize: 10,
  realSize: 0,
  list: [initRepository],
}

export const testPBIList: IBacklogItemList = {
  pageNumber: 1,
  pageCount: 1,
  pageSize: 10,
  realSize: 0,
  list: [initBI],
}

export const testSprintList: ISprintList = {
  pageNumber: 1,
  pageCount: 1,
  pageSize: 10,
  realSize: 0,
  list: [initSprint],
}

export const testTaskList: ITaskList = {
  pageNumber: 1,
  pageCount: 1,
  pageSize: 10,
  realSize: 0,
  list: [initTask],
}

export const tstConf = {
  ownerName: "ownerName",
  url: `${config.backend.ip}:${config.backend.port}/api`,
  pbiId: 0,
  sprintNr: 0,
  hours: 2,
  editedPBI: {
    "name": initBI.name,
    "priority": initBI.priority,
    "acceptanceCriteria": initBI.acceptanceCriteria
  },
  name: "taskName",
  taskId: 0,
  login: "userLogin",
  isAssign: true,
  hotfix: true
}

export const testConnectionError = {
  "code": 0,
  "response": {
    "data": null,
    "Message": "Connection error",
    "metadata": null,
    "successful": false,
  },
}
export const testFilters = { pageSize: config.defaultFilters.page, pageNumber: config.defaultFilters.size }

export const testFetchReposVals = {
  filters: testFilters,
  token: config.token,
}