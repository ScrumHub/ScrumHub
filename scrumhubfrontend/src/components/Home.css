
.reposGrid {
    display: grid;
    place-items: center;
    margin-top: 3vh;
    background: transparent;
  }
  
  .repoCard {
    background: white;
    margin-bottom: 3vh;
  }
  
  .calButton {
    width: 100%;
    overflow: hidden;
    white-space: nowrap;
  }

  .cardButton {
    width: 80%;
     /*width: 90%;*/
  }

  .ant-popover-disabled-compatible-wrapper{
      display:unset !important;
      width: 100%;
  }

  .ant-card-actions{
      padding-left: 24px !important;
      padding-right: 24px !important;
  }

