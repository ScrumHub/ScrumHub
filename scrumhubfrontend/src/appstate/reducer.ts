import * as Actions from "./actions";
import { createReducer, PayloadAction } from "@reduxjs/toolkit";
import { RequestResponse } from "./response";
import { IAssignBI, ILoginState, IPeopleList, IPerson, IBacklogItem, IBacklogItemList, IRepository, IRepositoryList, ISprint, ISprintList, ITask, ITaskList, IState } from "./stateInterfaces";
import { initError, loggedOutLoginState, unassignedPBI } from "./stateInitValues";
import { isArrayValid, isNameFilterValid } from "../components/utility/commonFunctions";
import { getError, updateStateTasks, updateStateKeys, addStateTask, addStateUnassignedTaskToPBI, updateStatePBI, updateTasksSWR, fetchStateRepos, updateStateOneSprint, addStateRepo, addStatePBI, addStateSprint, updateOnDragStateTasks } from "./stateUtilities";
import { isNull } from "lodash";
import config from "../configuration/config";
var _ = require('lodash');

export const reducerFunction = (initState: IState) => {
  return (createReducer(initState, {
    [Actions.clearError.type]: (state: IState, payload: any) => {
      let newState = _.cloneDeep(state as IState);
      return ({ ...newState, error: initError, changedRepo:isNameFilterValid(payload.payload)?payload.payload:""});
    },
    [Actions.clearReposList.type]: (state: IState) => {
      let newState = _.cloneDeep(state);
      newState.repositories = [];
      newState.reposRequireRefresh = true;
      newState.reposLastPage = false;
      return newState;
    },
    [Actions.clearState.type]: (state: IState) => {
      return (initState);
    },
    [Actions.clearProject.type]: (state: IState) => {
      let newState = _.cloneDeep(state) as IState;
      return { ...initState, repositories: newState.repositories, loginState:newState.loginState, currentUser: newState.currentUser };
    },
    [Actions.clearPBIsList.type]: (state: IState) => {
      let newState = _.cloneDeep(state);
      newState.pbiPage.list = [];
      newState.productRequireRefresh = true;
      newState.reposLastPage = false;
      return newState;
    },
    [Actions.clearSprintList.type]: (state: IState) => {
      let newState = _.cloneDeep(state);
      newState.sprintPage.list = [];
      newState.sprintRequireRefresh = true;
      newState.sprintLastPage = false;
      return newState;
    },
    [Actions.clearSprint.type]: (state: IState) => {
      let newState = _.cloneDeep(state);
      newState.openSprint = null;
      newState.sprintRequireRefresh = true;
      newState.sprintLastPage = false;
      return newState;
    },
    [Actions.updateSprintKeys.type]: (state: IState, expanded: any) => {
      let newState = _.cloneDeep(state);
      const temp = expanded.payload as number[];
      return { ...newState, keys: { ...newState.keys, sprintKeys: updateStateKeys(newState.keys.sprintKeys, temp) } };
    },
    [Actions.updatePBIKeys.type]: (state: IState, expanded: any) => {
      let newState = _.cloneDeep(state);
      const temp = expanded.payload as number[];
      return { ...newState, keys: { ...newState.keys, pbiKeys: updateStateKeys(newState.keys.pbiKeys, temp) } };
    },
    [Actions.updateSprintLoadingKeys.type]: (state: IState, expanded: any) => {
      let newState = _.cloneDeep(state);
      const temp = expanded.payload as number[];
      return { ...newState, loadingKeys: { ...newState.loadingKeys, sprintKeys: updateStateKeys(newState.loadingKeys.sprintKeys, temp) } };
    },
    [Actions.updatePBILoadingKeys.type]: (state: IState, expanded: any) => {
      let newState = _.cloneDeep(state);
      const temp = expanded.payload as number[];
      return { ...newState, loadingKeys: { ...newState.loadingKeys, pbiKeys: updateStateKeys(newState.loadingKeys.pbiKeys, temp) } };
    },
    [Actions.login.type]: (state: IState, expanded: any) => {
      let newState = _.cloneDeep(state);
      const tempLogin = expanded.payload as ILoginState
      return ({ ...newState, loginState: { ...newState.loginState, isLoggedIn: tempLogin.isLoggedIn, token: tempLogin.token } });
    },
    [Actions.logout.type]: (state: IState) => { return ({ ...initState, loginState: loggedOutLoginState }); },
    /*REPOS*/
    [Actions.fetchReposThunk.pending.toString()]: (state: IState) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.fetchReposThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IRepositoryList, number>>) => {
      let newState = _.cloneDeep(state);
      return fetchStateRepos(newState, payload.payload.response as IRepositoryList, _.get(payload, ["meta", "arg", "filters", "pageNumber"], config.defaultFilters.page),
       _.get(payload, ["meta", "arg", "filters", "pageSize"], config.defaultFilters.size),_.get(payload.payload.response, ["pageCount"], 1));
    },
    [Actions.fetchReposThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.addRepositoryThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.addRepositoryThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IRepository, number>>) => {
      let newState = _.cloneDeep(state);
      return (addStateRepo(newState, payload.payload.response as IRepository));
    },
    [Actions.addRepositoryThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },/*PBIS*/
    [Actions.finishPBIThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.finishPBIThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IBacklogItem, number>>) => {
      let newState = _.cloneDeep(state);
      return updateStatePBI(newState, payload.payload.response as IBacklogItem);
    },
    [Actions.finishPBIThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.addPBIThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.addPBIThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IBacklogItem, number>>) => {
      let newState = _.cloneDeep(state);
      return(addStatePBI(newState,payload.payload.response as IBacklogItem));
    },
    [Actions.addPBIThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.deletePBIThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.deletePBIThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<any, any>>) => {
      let newState = _.cloneDeep(state);
      newState.loading = false;
      return newState;
    },
    [Actions.deletePBIThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.fetchPBIsThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.fetchPBIsThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IBacklogItemList, number>>) => {
      let newState = _.cloneDeep(state);
      const pbisList = payload.payload.response as IBacklogItemList;
      if (newState.pbiPage && isArrayValid(newState.pbiPage.list) && newState.pbiPage.list[0].id === 0) {
        newState.pbiPage = { ...pbisList, list: [newState.pbiPage.list[0]].concat(pbisList.list) };
      }
      else {
        newState.pbiPage = { ...pbisList, list: [unassignedPBI].concat(pbisList.list) };
      }
      newState.loading = false;
      return newState;
    },
    [Actions.fetchPBIsThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.fetchPeopleThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.fetchPeopleThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IPeopleList, number>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, people: payload.payload.response as IPeopleList, error: initError, loading: false };
    },
    [Actions.fetchPeopleThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.getCurrentUserThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.getCurrentUserThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IPerson, number>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, currentUser: payload.payload.response as IPerson, error: initError, loading: false };
    },
    [Actions.getCurrentUserThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.getPBINamesThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.getPBINamesThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IBacklogItemList, number>>) => {
      let newState = _.cloneDeep(state);
      newState.error = initError;
      newState.namedPBI = (payload.payload.response as IBacklogItemList).list as IAssignBI[];
      newState.loading = false;
      return newState;
    },
    [Actions.getPBINamesThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.estimatePBIThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.estimatePBIThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IBacklogItem, number>>) => {
      let newState = _.cloneDeep(state);
      return updateStatePBI(newState, payload.payload.response as IBacklogItem);
    },
    [Actions.estimatePBIThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.editPBIThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.editPBIThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IBacklogItem, number>>) => {
      let newState = _.cloneDeep(state);
      return updateStatePBI(newState, payload.payload.response as IBacklogItem);
    },
    [Actions.editPBIThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },/*SPRINTS*/
    [Actions.fetchSprintsThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.fetchSprintsThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<ISprintList, number>>) => {
      let newState = _.cloneDeep(state);
      const sprints = payload.payload.response as ISprintList;
      newState.sprintPage = sprints;
      const index = sprints && isArrayValid(sprints.list) ? sprints.list.findIndex((sprint: ISprint) => sprint.isCurrent) : -1;
      if (index !== -1) { newState.activeSprintNumber = sprints.list[index].sprintNumber };
      newState.sprintRequireRefresh = false;
      newState.loading = false;
      newState.error = initError;
      return newState;
    },
    [Actions.fetchSprintsThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.fetchOneSprintThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.fetchOneSprintThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<ISprint, number>>) => {
      let newState = _.cloneDeep(state);
      newState.openSprint = payload.payload.response as ISprint;
      if (!isNull(newState.openSprint) && newState.openSprint.isCurrent) { newState.activeSprintNumber = newState.openSprint.sprintNumber };
      newState.sprintRequireRefresh = false;
      newState.error = initError;
      newState.loading = false;
      return newState;
    },
    [Actions.fetchOneSprintThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.updateOneSprintThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.updateOneSprintThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<ISprint | any, number>>) => {
      let newState = _.cloneDeep(state);
      return (updateStateOneSprint(newState, payload.payload.response as ISprint));
    },
    [Actions.updateOneSprintThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.completeOneSprintThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.completeOneSprintThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<ISprint, number>>) => {
      let newState = _.cloneDeep(state);
      return (updateStateOneSprint(newState, payload.payload.response as ISprint));
    },
    [Actions.completeOneSprintThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.addSprintThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.addSprintThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<ISprint, undefined>>) => {
      let newState = _.cloneDeep(state);
      return(addStateSprint(newState,payload.payload.response as ISprint));
    },
    [Actions.addSprintThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },/*TASKS*/
    [Actions.fetchPBITasksThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.fetchPBITasksThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<ITaskList, number>>) => {
      let newState = _.cloneDeep(state);
      return updateTasksSWR(newState, payload.payload.response as ITaskList);
    },
    [Actions.updateTasks.type]: (state: IState, payload: PayloadAction<any>) => {
      let newState = _.cloneDeep(state);
      const item = payload.payload as IBacklogItem;
      return updateStatePBI(newState, item);
    },
    [Actions.fetchPBITasksThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.addUnassignedTasksToPBI.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.addUnassignedTasksToPBI.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<ITaskList, number>>) => {
      let newState = _.cloneDeep(state);
      return addStateUnassignedTaskToPBI(newState, payload.payload.response as ITaskList);
    },
    [Actions.addUnassignedTasksToPBI.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.addTaskThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.addTaskThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<ITask, number>>) => {
      let newState = _.cloneDeep(state);
      return addStateTask(newState, payload.payload.response as ITask);
    },
    [Actions.addTaskThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.assignTaskToPBIThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.assignTaskToPBIThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<IBacklogItem, number>>) => {
      let newState = _.cloneDeep(state);
      return updateOnDragStateTasks(newState,payload.payload.response as ITask);
    },
    [Actions.assignTaskToPBIThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.updatePersonInTaskThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.updatePersonInTaskThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<ITask, number>>) => {
      let newState = _.cloneDeep(state);
      return updateStateTasks(newState, payload.payload.response as ITask);
    },
    [Actions.updatePersonInTaskThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
    [Actions.startTaskThunk.pending.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, loading: true };
    },
    [Actions.startTaskThunk.fulfilled.toString()]: (state: IState, payload: PayloadAction<RequestResponse<ITask, number>>) => {
      let newState = _.cloneDeep(state);
      return updateStateTasks(newState, payload.payload.response as ITask);
    },
    [Actions.startTaskThunk.rejected.toString()]: (state: IState, payload: PayloadAction<RequestResponse<undefined, undefined>>) => {
      let newState = _.cloneDeep(state);
      return { ...newState, error: getError(payload.payload), loading: false };
    },
  }));
}