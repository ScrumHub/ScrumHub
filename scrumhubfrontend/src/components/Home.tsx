import { useEffect, useState } from 'react';
import { Button, Card } from 'antd';
import * as Actions from '../appstate/actions';
import 'antd/dist/antd.css';
import './Home.css';
import { IFilters, IRepository, IState } from '../appstate/stateInterfaces';
import { useNavigate } from 'react-router';
import InfiniteScroll from 'react-infinite-scroll-component';
import config from '../configuration/config';
import { useSelector } from 'react-redux';
import { CalendarOutlined, FolderAddOutlined, InfoCircleOutlined } from '@ant-design/icons';
import { store } from '../appstate/store';
import { clearReposList } from '../appstate/actions';
import SkeletonList, { CantAddToShButton, InShButton } from './utility/LoadAnimations';
import { dateFormat, isArrayValid } from './utility/commonFunctions';
const { Meta } = Card;


export function Home() {
  const isLoggedIn = useSelector((appState: IState) => appState.loginState.isLoggedIn);
  const token = useSelector((appState: IState) => appState.loginState.token);
  const [filters, setFilters] = useState<IFilters>({pageNumber: config.defaultFilters.page,pageSize: config.defaultFilters.size,});
  const lastPage = useSelector((state: IState) => state.reposLastPage);
  const [displayId, setDisplayId] = useState(-1);
  const [fetching, setFetching] = useState(false);
  const loading = useSelector((state: IState) => state.loading);
  const refreshRequired = useSelector(
    (appState: IState) => appState.reposRequireRefresh as boolean
  );
  const repos = useSelector(
    (state: IState) => state.repositories as IRepository[]
  );
  const navigate = useNavigate();
  const [initialRefresh, setInitialRefresh] = useState(true);
  useEffect(() => {
    if (initialRefresh) {
      setInitialRefresh(false);
      store.dispatch(clearReposList());
    }
  }, [initialRefresh]);

  useEffect(() => {
    if (isLoggedIn && refreshRequired) {
      try {
        store.dispatch(
          Actions.fetchRepositoriesThunk({
            filters: {
              pageSize: config.defaultFilters.size,
              pageNumber: config.defaultFilters.page,
            },
            token: token,
          })
        );
      } catch (err) {
        console.error("Failed to fetch the repos: ", err);
      }
      finally {
        setFilters({ ...filters, pageNumber: config.defaultFilters.page + 1 });
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [refreshRequired, isLoggedIn]);

  const fetchMore = () => {
    if (!fetching) {
      setFetching(true);
      try {
        store.dispatch(
          Actions.fetchRepositoriesThunk({
            filters: {
              ...filters,
              pageNumber: filters.pageNumber,
            },
            token: token,
          }) 
        );
      } catch (err) {
        console.error("Failed to fetch the repos: ", err);
      } finally {
        setFilters({ ...filters, pageNumber: filters.pageNumber + 1 });
        setFetching(false);
      }
    }
  };

  function addProject(projId: number) {
    setDisplayId(projId);
      store.dispatch(
        Actions.addRepositoryThunk({
          id: projId,
          token: token,
        })
      )
      .catch((error: any)=>{console.error("Failed to fetch the pbis: ", error);
      setDisplayId(-1);
      return({});})
      .then(()=>setDisplayId(-1));
  };

  function redirectToProject(props: IRepository) {
    localStorage.setItem("ownerName", props.name);
    navigate(`/${props.name.split("/")[0]}/${props.name.split("/")[1]}`, { replace: true });
  };

  return (
    <div>
      <InfiniteScroll
      scrollableTarget="scrollableDiv"
      dataLength={isArrayValid(repos) ? repos.length : 0}
      scrollThreshold={0.7}
      next={fetchMore}
      hasMore={!lastPage && !fetching}
      loader={<></>}>
      <div className="reposGrid">
        {isArrayValid(repos) && repos.map((rep: IRepository) => {
          return (<section className="card" style={{ width: "85%" }} key={rep.gitHubId} >
            <Card loading={rep.gitHubId===displayId &&loading } className='repoCard' type="inner" actions=
              {[!rep.hasAdminRights ? <CantAddToShButton/>:(rep.alreadyInScrumHub ? <InShButton/>:
                <Button loading={rep.gitHubId===displayId &&loading } disabled={false} className='cardButton' onClick={() => { addProject(rep.gitHubId);}} >
                <span>{<FolderAddOutlined disabled={false} />}{" Add to ScrumHub" }</span></Button>),
              <Button loading={rep.gitHubId===displayId &&loading} disabled={!rep.alreadyInScrumHub} type="primary" className='cardButton' onClick={() => { redirectToProject(rep) }}><span><InfoCircleOutlined />{" Manage project"}</span></Button>,
              <Button className='calButton'><span><CalendarOutlined />{rep.typeOfLastActivity === "No recent activity" ? " Not updated" : " " + dateFormat(rep.dateOfLastActivity as Date)}</span></Button>]}>
              <Meta
                title={rep.alreadyInScrumHub ? <div className='link-button' onClick={() => { redirectToProject(rep) }}>{rep.name.split("/")[1]}</div> : <div>{rep.name.split("/")[1]}</div>}
                description={rep.description}
              ></Meta>
              <br />
              <p>{"This project has " + (rep.sprints && rep.sprints.length === 1 ? "1 sprint and " : (rep.sprints.length + " sprints and ")) + (rep.backlogItems && rep.backlogItems.length === 1 ? "1 backlog item.\n" : rep.backlogItems.length + " backlog items.\n")}</p>
              <p>{rep.typeOfLastActivity + (rep.dateOfLastActivity === "No recent activity" ? " " : " was the last activity") + " in the repository."}</p>
            </Card>
          </section>);
        })
        }
        <SkeletonList loading={loading  || refreshRequired} number={repos.length < 1?5:1} />
      </div>
    </InfiniteScroll>
    </div >
  );
}