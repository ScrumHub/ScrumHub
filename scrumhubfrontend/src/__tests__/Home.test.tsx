/**
 * @jest-environment jsdom
 */

import React from "react";
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";
import { isNull } from "lodash";
import { Provider } from "react-redux";
import { Route, BrowserRouter as Router, Routes } from "react-router-dom";
import { configureStore } from "@reduxjs/toolkit";
import { reducerFunction } from "../appstate/reducer";
import { initTestState } from "../appstate/stateTestValues";
import renderer from 'react-test-renderer';
import { Home } from "../components/Home";

describe('Home component succesfully rendered', () => {
  let container: any;
  let store: any;
  beforeEach(() => {
    container = document.createElement('div');
    document.body.appendChild(container);
    store = configureStore({
      reducer: reducerFunction(initTestState),
    });
  });

  afterEach(() => {
    // cleanup on exiting
    if (!isNull(container)) {
      unmountComponentAtNode(container);
      container.remove();
      container = null;
    }
  });

  it("Rendered list of repos", async () => {
    await act(async () => {
      render(
        <Provider store={store}>
          <Router>
            <Routes>
              <Route path="/" element={<Home />} />
            </Routes>
          </Router>
        </Provider>,
        container
      );
    });
    expect(
      container?.getElementsByClassName("container").length
    ).toBeGreaterThanOrEqual(0);
    expect(container.textContent).toBe("");
  });

  it('home component rendered is the same as snapshot', () => {
    const temp = renderer.create(
      <Provider store={store}>
        <Router>
          <Routes>
            <Route path="/" element={<Home />} />
          </Routes>
        </Router>
      </Provider>).toJSON();
    expect(temp).toMatchSnapshot();
  });
});
