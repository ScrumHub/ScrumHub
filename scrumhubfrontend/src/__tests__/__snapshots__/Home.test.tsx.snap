// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home component succesfully rendered home component rendered is the same as snapshot 1`] = `
<div>
  <div
    className="infinite-scroll-component__outerdiv"
    style={Object {}}
  >
    <div
      className="infinite-scroll-component "
      style={
        Object {
          "WebkitOverflowScrolling": "touch",
          "height": "auto",
          "overflow": "auto",
        }
      }
    >
      <div
        className="reposGrid"
      >
        <section
          className="card"
          style={
            Object {
              "width": "85%",
            }
          }
        />
      </div>
    </div>
  </div>
</div>
`;
